ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/ForeignKeyJoinParamsFactory.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams;
+
+import io.confluent.ksql.name.ColumnName;
+import io.confluent.ksql.schema.ksql.Column;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.LogicalSchema.Builder;
+import java.util.Optional;
+
+public final class ForeignKeyJoinParamsFactory {
+
+  private ForeignKeyJoinParamsFactory() {
+  }
+
+  public static <KRightT> ForeignKeyJoinParams<KRightT> create(
+      final ColumnName leftJoinColumnName,
+      final LogicalSchema leftSchema,
+      final LogicalSchema rightSchema
+  ) {
+    return new ForeignKeyJoinParams<>(
+        createKeyExtractor(leftSchema, leftJoinColumnName),
+        new KsqlValueJoiner(leftSchema.value().size(), rightSchema.value().size(), 0),
+        createSchema(leftSchema, rightSchema)
+    );
+  }
+
+  public static LogicalSchema createSchema(
+      final LogicalSchema leftSchema,
+      final LogicalSchema rightSchema
+  ) {
```

Comment:
We need to validate that the join expression types match. If we follow the pattern of the existing JoinParamsFactory, we can do that here. 

JoinParamsFactory also validates that the join expressions represent the complete key, which is a requirement for the right side of FK joins as well. I think we don't need this extra validation here in light of the recent updates to LogicalPlanner to check this condition up front, but it may make sense to have a duplicate check as a second guard in the physical layer here as well.

Reply:
> We need to validate that the join expression types match. If we follow the pattern of the existing JoinParamsFactory, we can do that here.

Yes, but this should be done in the logical planner IMHO.

Happy to add a second guard.

Reply:
Where are we currently validating that the expression types match? I don't see the check in either the logical or the physical layer right now.

Reply:
For table-table joins, it happens in `JoinParamsFactory.throwOnKeyMismatch` (but I think it's wrong to do this check when building the physical plan, but it should be done earlier.

Thus, I did not add it to `ForeignKeyJoinParamsFactory` and I think @spena current WIP PR should add this check. (And we should move the existing check for PK-joins in a follow up PR into the logical planner.)

Reply:
`JoinParamsFactory.throwOnKeyMismatch()` is called from `JoinParamsFactory.createSchema()`, which is called from `JoinNode.buildJoinSchema()`, which is called as part of attempting to build the join from the logical planner. In other words, this check _is_ already happening in the logical planner (for table-table joins today). Where are you proposing that the check should happen instead?

Reply:
That is quite convoluted code than, because `JoinParamsFactory` is in package `io.confluent.ksql.execution.streams` -- the logical planner should not use helpers from the physical planner IMHO. Maybe the solution is to move the whole factory class into the logical planner package?

Reply:
The method is also called from the physical layer when executing the execution step. Ideally this method would be in a separate package clearly meant to be shared between the two, rather than entirely in one or the other.

Reply:
Interesting. Think we don't need to resolve this in this PR. We only need to make sure that @spena PR's adds the type check.